### ISSUE TITLE
Track GHC commit

### ISSUE THREAD

**Issue by izgzhen** [snowleopard/hadrian#306]
It looks like there is no way to associate a particular hadrian commit to a GHC commit, so the build failure caused by potential unexpected GHC update like in #305 can be very hard to detect.

I suggest tagging the git hash of the most recent successfully built GHC somewhere, and instruct user/CI to check out corresponding version automatically.

**Comment by ndmitchell** [comment]
Does Hadrian have a nightly cron build? That's now an option in Travis and might detect GHC changes easily and quickly.

**Comment by snowleopard** [comment]
@izgzhen I don't think tagging will help much in practice: we're pretty much always in the situation that the last Hadrian commit is broken by one of a new commits in GHC. The tag will only give us the lower bound, but it can also be obtained by looking at the commit time.

@ndmitchell We don't at the moment. I think regular builds is a better solution compared to tagging. 

I'll look into enabling cron builds on Travis.

**Comment by izgzhen** [comment]
> but it can also be obtained by looking at the commit time

@snowleopard I am not sure this will always be true. If we can't promise to reject all commits that failed some build, there will be a mismatch. Also, looking at commit time is not very convenient & precise.

**Comment by ndmitchell** [comment]
@snowleopard - it's pretty trivial now. Let me know if you get stuck.

@izgzhen provided you aren't unlucky and commit within 24 of a rare upstream breakage you should be fine - and even if you do something else will break quickly. I often wish that on a failed pull request Travis did a sanity build of the original.

**Comment by snowleopard** [comment]
@ndmitchell I easily found the setting in Travis, so hopefully we will now have daily builds. Thanks for the hint! (Was travelling all week so couldn't do this earlier.)

**Comment by snowleopard** [comment]
@izgzhen I don't quite understand how you propose to make tagging work. 

Imagine I've have made a patch to Hadrian which works on my Windows machine. I push it to the repo and it fails on the Mac OSX instance. What do I do now? Do I tag as "For GHC commit N, this works on Windows & Linux, but fails on Mac OS X"? Is this really helpful? To me this seems to introduce extra work/mental overhead to the development process without substantial benefit.

**Comment by izgzhen** [comment]
> Is this really helpful? To me this seems to introduce extra work/mental overhead to the development process without substantial benefit

@snowleopard First what you described is basically what I have in mind: a precise, 100% reproducible history of builds. But yes, you have a point in that it might lead to extra overhead since we don't have a infra person for automating all the related bits now :)

You maybe close this issue if you see fit -- I guess if we ever ship Hadrian into GHC tree, then it is GHC team's job to make sure everything is well synchronized.

**Comment by snowleopard** [comment]
@izgzhen I see. It's a useful goal, but I agree that it will be much easier to achieve once Hadrian is merged into the GHC tree, so I'm closing this.

P.S.: I'm finally getting some free time to look into Hadrian, so I'll try to fix the build during this week. I've already fixed the Windows build, so we have the green light from AppVeyor. Will look into the recursive rules for `iserv` next.
